apps/src/javalab/javalabRedux.js

```diff
@@ -32,17 +33,45 @@ const CLOSE_PHOTO_PROMPTER = 'javalab/CLOSE_PHOTO_PROMPTER';
 const SET_IS_READONLY_WORKSPACE = 'javalab/SET_IS_READONLY_WORKSPACE';
 const SET_HAS_OPEN_CODE_REVIEW = 'javalab/SET_HAS_OPEN_CODE_REVIEW';
 const SET_COMMIT_SAVE_STATUS = 'javalab/SET_COMMIT_SAVE_STATUS';
+const SET_EDIT_TAB_KEY = 'javalab/SET_EDIT_TAB_KEY';
+const SET_ACTIVE_TAB_KEY = 'javalab/SET_ACTIVE_TAB_KEY';
+const SET_FILE_METADATA = 'javalab/SET_FILE_METADATA';
+const SET_ORDERED_TAB_KEYS = 'javalab/SET_ORDERED_TAB_KEYS';
+const SET_ALL_EDITOR_METADATA = 'javalab/SET_EDITOR_METADATA';
+
+export const getTabKey = index => `file-${index}`;
+
+const fileMetadataForEditor = (sources, isEditingStartSources) => {
```

Comment:
could this logic of converting sources -> fileMetadata be done in some helper function elsewhere (that is called before setting fileMetadata in redux)? Generally I think we want to keep redux free of complex logic.

Reply:
Ok makes sense -- I stumbled across this file yesterday, might be a good place for it:

https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/javalab/JavalabFileHelper.js